### PR TITLE
[imagecache] Add a process to clean old images from cache

### PR DESCRIPTION
**Pull request by rmrector**
## Description
Adds a conservative image cache cleaning process. It removes images from the cache if they are not in the video or music library and have not been used in the past month.

Also checks the add-on database for images. Ignores PVR images as those cached images are managed another way. Implemented as a background process checking up to 1000 images up to 4 times per day.

When checking the video library it considers [the "Artwork level" settings](https://kodi.wiki/view/Settings/Media/Videos#Artwork). Same for the [music library](https://kodi.wiki/view/Settings/Media/Music#Artwork) except treats "Custom" like Maximum, since it has a couple of special rules.

PR also adds an action in settings to clear all old images at once.

## Motivation and context
Some devices don't have enough storage space to hold every image Kodi ever displayed, and it shouldn't even if there is enough space.

These days the texture cache is likely to take up a relatively large amount of device storage space, this is one conservative step to limiting that amount.

## How has this been tested?
Manual testing on Windows and Pi3 and dogfooding on Pi5 for a couple weeks.

## What is the effect on users?
Prevent the image cache from growing indefinitely. Solve out-of-space issues when library artwork fits in 50% - 80% of storage space, from CCwGTV and tiny library / just streaming to 16GB storage and large libraries with a lot of artwork.

## Screenshots (if appropriate):
image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by howie-f** [comment]
cool feature. will my actor thumbs survive the purge? those that were scraped in long ago, then got cached and are no longer available at the location they were originally pulled in from?

**Comment by rmrector** [comment]
> cool feature. will my actor thumbs survive the purge? those that were scraped in long ago, then got cached and are no longer available at the location they were originally pulled in from?

yes, it keeps images in the video library.

**Comment by Hitcher** [comment]
Sounds like a really useful feature. Would it also be possible to add a manual clean setting somewhere in Kodi?

Thanks.

**Comment by zag2me** [comment]
Superb addition. Will really help people who use the YouTube Add-on I think. My Shield is always running out of space.

**Comment by rmrector** [comment]
> Would it also be possible to add a manual clean setting somewhere in Kodi?

Yes, that seems reasonable. It could check all images not used in the past 30 days.

> Will really help people who use the YouTube Add-on I think.

Ya, I figure YouTube in particular can accumulate a lot of images.

**Comment by mkreisl** [comment]
I integrated the PR all the time. But as of today it's full of crap and only produces compile errors.

Built under Debian Bullseye

**Comment by thexai** [comment]
Nice feature. Tested on Android (Shield) and works as expected:

**Before:**
before

**After:**
after

Clean is really fast. No issues found.

**Review by thexai — APPROVED** [review]
Code looks good

**Comment by rmrector** [comment]
Thanks!

**Comment by rmrector** [comment]
I added a quick mention to the wiki on the [Artwork/Cache](https://kodi.wiki/view/Artwork/Cache#Automatic_cache_cleaning) page.

File: xbmc/video/VideoDatabase.cpp [hunk 1]
```diff
@@ -12750,3 +12751,131 @@ void CVideoDatabase::SetVideoVersionDefaultArt(int dbId, int idFrom, VideoDbCont
...
+    // then check any chapter thumbnails against path and file tables
+
+    if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+            CSettings::SETTING_MYVIDEOS_EXTRACTCHAPTERTHUMBS))
+    {
+      std::vector<std::string> foundVideoFiles;
+      for (const auto& image : imagesToCheck)
+      {
+        auto imageFile = IMAGE_FILES::CImageFileURL(image);
+        if (imageFile.GetSpecialType() == "video" && !imageFile.GetOption("chapter").empty())
```
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:12853]: Does this include user and episode bookmarks (thumNailImage url stored in bookmark table, type 0 and 2)?

Comment (rmrector) [reply xbmc/video/VideoDatabase.cpp:12853]: It does not.

Bookmarks are generated straight to files at "userdata/Thumbnails/Video/Bookmarks/", outside of the texture cache system. The resulting images shouldn't be saved in the texture cache either - the files already exist nearby, on most likely the same filesystem.

**Pull request merged**